### PR TITLE
[WIP][CORE] Optimize shuffle partitioner algorithm to take advantage of  auto-vectorization in GCC

### PR DESCRIPTION
**Pull request by kecookier**
## What changes were proposed in this pull request?
Why we make this change?

We made vs. benchmark in https://quick-bench.com/, And the result shows the speed of the improved Partitioner::computer is 2x faster than the old implementation.

(Fixes: \#ISSUE-ID)

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

### CODE REVIEW

File: cpp/core/shuffle/RoundRobinPartitioner.cc [hunk 1]
```diff
@@ -22,15 +22,27 @@ namespace gluten {
...
-  for (auto& pid : partitionId) {
-    pid = pidSelection_;
-    partitionIdCnt[pidSelection_]++;
-    pidSelection_ = (pidSelection_ + 1) == numPartitions_ ? 0 : (pidSelection_ + 1);
+    std::vector<uint16_t>& row2Partition,
+    std::vector<uint32_t>& partition2RowCount) {
+  std::fill(std::begin(partition2RowCount), std::end(partition2RowCount), 0);
+  row2Partition.resize(numRows);
+
+  int32_t pidSelection = 0;
```
Comment (zuochunwei) [inline cpp/core/shuffle/RoundRobinPartitioner.cc:30]: We should use `pidSelection_` here.
You can use a local variable `int pidSelection` initialized from `pidSelection_` in the following loop, and store it into `pidSelection_` after the loop

File: cpp/core/shuffle/RoundRobinPartitioner.h [hunk 2]
```diff
@@ -27,11 +27,8 @@ class RoundRobinPartitioner final : public ShuffleWriter::Partitioner {
   arrow::Status compute(
       const int32_t* pidArr,
       const int64_t numRows,
-      std::vector<uint16_t>& partitionId,
-      std::vector<uint32_t>& partitionIdCnt) override;
-
- private:
-  int32_t pidSelection_ = 0;
```
Comment (zuochunwei) [inline cpp/core/shuffle/RoundRobinPartitioner.h:34]: don't delete it, it's used to save the status

**Comment by zuochunwei** [comment]
supply some information about why we are making this change,  We made vs. benchmark in `https://quick-bench.com/`, and the result shows the speed of the improved `Partitioner::computer` is 2x faster than the old implement.

File: cpp/core/shuffle/RoundRobinPartitioner.cc (hunk 1 above)
Comment (kecookier) [reply cpp/core/shuffle/RoundRobinPartitioner.cc:30]: ok

**Review by zuochunwei — APPROVED** [review]

**Comment by PHILO-HE** [comment]
Just a small issue in merging code. Let's remove `[WIP]` in final commit message next time. Thanks!

**Pull request merged**